chain/chain/src/update_shard.rs

```diff
@@ -45,8 +45,20 @@ pub struct StateSplitResult {
     pub(crate) results: Vec<ApplySplitStateResult>,
 }
 
+/// Result of processing shard update, covering both stateful and stateless scenarios.
 #[derive(Debug)]
-pub enum ApplyChunkResult {
+pub enum ShardUpdateResult {
+    /// Stateful scenario - processed update for a single block.
+    Stateful(ShardBlockUpdateResult),
+    /// Stateless scenario - processed update based on state witness in a chunk.
+    /// Contains `ChunkExtra`s - results for processing updates corresponding
+    /// to state witness.
+    Stateless(Vec<(CryptoHash, ShardUId, ChunkExtra)>),
+}
+
+/// Result for a shard update for a single block.
+#[derive(Debug)]
+pub enum ShardBlockUpdateResult {
```

Comment:
nit: maybe `ShardChunkUpdateResult` would be more precise

Reply:
I chose `ShardBlockUpdateResult` explicitly because chunk is not necessarily present. Especially if there is a state split. So I'll bet on that for now